gcsio/src/main/java/com/google/cloud/hadoop/gcsio/StorageProvider.java

```diff
@@ -0,0 +1,196 @@
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Provides GCS SDK Storage object which is used to access the GCS. Also Caches the storage objects,
+ * so they can be used re-used.
+ */
+public class StorageProvider {
+
+  // TODO: Replace Storage with a StorageWrapper which does not expose the close method. The
+  //  dependants of this provider should not be able to close the storage accidentally. It should
+  //  always be managed through this provider.
+  @VisibleForTesting
+  final Cache<StorageProviderCacheKey, Storage> cache =
+      CacheBuilder.newBuilder().recordStats().build();
+
+  /**
+   * Tracks the number of times a storage client is used. Used to determine when a storage can be
+   * closed.
+   */
+  @VisibleForTesting
+  final Map<Storage, Integer> storageClientToReferenceMap = new ConcurrentHashMap<>();
+
+  /** Reverse map for storage reference to cache keys. */
+  final Map<Storage, StorageProviderCacheKey> storageToCacheKeyMap = new ConcurrentHashMap<>();
+
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  @VisibleForTesting
+  StorageProviderCacheKey computeCacheKey(
+      Credentials credentials,
+      GoogleCloudStorageOptions storageOptions,
+      Function<List<AccessBoundary>, String> downscopedAccessTokenFn) {
+    return StorageProviderCacheKey.builder()
+        .setCredentials(credentials)
+        .setHttpHeaders(storageOptions.getHttpRequestHeaders())
+        .setIsDirectPathPreferred(storageOptions.isDirectPathPreferred())
+        .setIsDownScopingEnabled(downscopedAccessTokenFn != null)
+        .setIsTracingEnabled(storageOptions.isTraceLogEnabled())
+        .setWriteChannelOptions(storageOptions.getWriteChannelOptions())
+        .setProjectId(storageOptions.getProjectId())
+        .build();
+  }
+
+  /** Determines if the storage instance can be served from the cache. */
+  private static boolean canCache(
+      GoogleCloudStorageOptions options,
+      List<ClientInterceptor> interceptors,
+      ExecutorService pCUExecutorService) {
+
+    return options.isStorageClientCachingExperimentEnabled()
+        // These values are currently passed while creating the storage client, but they are
+        // currently not configurable by the FS options so skipping caching.
+        && pCUExecutorService == null
+        && (interceptors == null || interceptors.isEmpty());
+  }
+
+  Storage getStorage(
+      Credentials credentials,
+      GoogleCloudStorageOptions storageOptions,
+      List<ClientInterceptor> interceptors,
+      ExecutorService pCUExecutorService,
+      Function<List<AccessBoundary>, String> downscopedAccessTokenFn)
+      throws IOException {
+    if (canCache(storageOptions, interceptors, pCUExecutorService)) {
+      StorageProviderCacheKey key =
+          computeCacheKey(credentials, storageOptions, downscopedAccessTokenFn);
+      Storage storage = cache.getIfPresent(key);
```

Comment:
Please see https://stackoverflow.com/a/70655578

Use `.cache.get(key, k -> calculateValueForKey(k))` on place of manual `getIfPresent` call.

Reply:
Used synchronized blocks as discussed offline.